Updated phase 1: shard block structure (#578)
ethereum/eth2.0-specs: specs/core/1_shard-data-chains.md

| `SHARD_PROPOSER_DOMAIN`| 129             |
| `SHARD_ATTESTER_DOMAIN`| 130             |

## Helper functions

#### get_split_offset

````python
def get_split_offset(list_size: int, chunks: int, index: int) -> int:
  """
  Returns a value such that for a list L, chunk count k and index i,
  split(L, k)[i] == L[get_split_offset(len(L), k, i): get_split_offset(len(L), k+1, i)]
  """
  return (len(list_size) * index) // chunks
````

#### get_persistent_committee

```python
def get_persistent_commmitee(seed: Bytes32,

>                             validators: List[Validator],

Reviewer: Fix indent of this line and the following two (add one space).

                            shard: ShardNumber,
                            epoch: EpochNumber) -> List[ValidatorIndex]:

>                   

Reviewer: add docstring

    later_validator_set = get_active_validator_indices(state.validators, later_committee_start)
    later_seed = generate_seed(state, later_committee_start)
    later_start_offset = get_split_offset(len(later_validator_set), SHARD_COUNT, shard)
    later_end_offset = get_split_offset(len(later_validator_set), SHARD_COUNT, shard+1)
    later_committee = [
        later_validator_set[get_permuted_index(i, len(later_validator_set), later_seed)]
        for i in range(later_start_offset, later_end_offset)
    ]
    
    def get_switchover_epoch(index):
        return (
            bytes_to_int(hash(earlier_seed + bytes3(index))[0:8]) %
            PERSISTENT_COMMITTEE_PERIOD
        )
       
    # Take not-yet-cycled-out validators from earlier committee and already-cycled-in validators from
    # later committee; return a sorted list of the union of the two, deduplicated
    return sorted(list(set(
        [i for i in earlier_committee if epoch % PERSISTENT_COMMITTEE_PERIOD < get_switchover_epoch(i)] +
        [i for i in later_committee if epoch % PERSISTENT_COMMITTEE_PERIOD >= get_switchover_epoch(i)]
    )))
```

> 

Author: ```suggestion
#### get_shard_proposer_index

```python
def get_shard_proposer_index(state: BeaconState, shard: ShardNumber, slot: SlotNumber) -> ValidatorIndex:
    seed = hash(state.current_epoch_seed +
                int_to_bytes8(shard) +
                int_to_bytes8(slot))
    persistent_committee = get_persistent_committee(state, shard, slot_to_epoch(slot))
    return persistent_committee[bytes_to_int(seed[0:8]) % len(validators)]
```
```

Reviewer: looks good

* Verify that `beacon_chain_ref` is equal to or a descendant of the `beacon_chain_ref` specified in the `ShardBlock` pointed to by `parent_root`.
* Let `state` be the state of the beacon chain block referred to by `beacon_chain_ref`. Let `persistent_committee` be `[persistent_committee[i] for i in get_persistent_committee(state, slot, shard_id)`.
* Assert `verify_bitfield(participation_bitfield, len(persistent_committee))`
* Let `proposer_index = hash(state.randao_mix + int_to_bytes8(shard_id) + int_to_bytes8(slot)) % len(validators)`. Let `msg` be the block but with the `block.signature` set to `[0, 0]`. Verify that `BLSVerify(pub=validators[proposer_index].pubkey, msg=hash(msg), sig=block.signature, domain=get_domain(state, slot, SHARD_PROPOSER_DOMAIN))` passes.
* Let `group_public_key = bls_aggregate_pubkeys([state.validators[index].pubkey for i, index in enumerate(persistent_committee) if get_bitfield_bit(participation_bitfield, i) is True])`. Verify that `bls_verify(pubkey=group_public_key, msg=parent_root, sig=block.aggregate_signature, domain=get_domain(state, slot, SHARD_ATTESTER_DOMAIN))` passes.

We define the helper `get_proposal_committee` as follows:

```python

> def get_proposal_committee(seed: Bytes32,

Reviewer: Move this function up to a "helpers" section. This is important enough to pull it out from the block processing section.

> def get_proposal_committee(seed: Bytes32,

Author: Let's do persistent.

(edited all the appearances to make it that way)

                          validators: List[Validator],

>                           shard: int,

Reviewer: ```suggestion
                          shard: ShardNumber,
```

                          epoch: EpochNumber) -> List[ValidatorIndex]:
                  
    earlier_committee_start = epoch - (epoch % PROPOSAL_RESHUFFLE_PERIOD) - PROPOSAL_RESHUFFLE_PERIOD * 2

>     earlier_committee = split(shuffle(

Reviewer 2: `shuffle` function has been replaced with `get_permuted_index` style shuffling.

        get_active_validator_indices(validators, earlier_committee_start),
        generate_seed(state, earlier_committee_start)
    ), SHARD_COUNT)[shard]
    
    later_committee_start = epoch - (epoch % PROPOSAL_RESHUFFLE_PERIOD) - PROPOSAL_RESHUFFLE_PERIOD

>     later_committee = split(shuffle(

Reviewer 2: ditto

>         get_active_validator_indices(validators, earlier_committee_start),

Reviewer: ```suggestion
        get_active_validator_indices(validators, later_committee_start),
```

>         generate_seed(state, earlier_committee_start)

Reviewer: ```suggestion
        generate_seed(state, later_committee_start)
```

    ), SHARD_COUNT)[shard]
    
    def get_switchover_epoch(index):
        return (

>             int.from_bytes(hash(generate_seed(state, earlier_committee_start) + bytes3(index)), 'little') %

Reviewer: ```suggestion
            bytes_to_int(hash(generate_seed(state, earlier_committee_start) + bytes3(index))[0:8]) %
```

    'beacon_chain_ref': 'bytes32',
    # Merkle root of data
    'data_root': 'bytes32'
    # State root (placeholder for now)
    'state_root': 'bytes32',
    # Block signature
    'signature': 'bytes96',
    # Attestation
    'participation_bitfield': 'bytes',
    'aggregate_signature': 'bytes96',
}
```

## Shard block processing

For a `shard_block` on a shard to be processed by a node, the following conditions must be met:

* The `ShardBlock` pointed to by `shard_block.parent_root` has already been processed and accepted
* The signature for the block from the _proposer_ (see below for definition) of that block is included along with the block in the network message object

To validate a block header on shard `shard_block.shard_id`, compute as follows:

* Verify that `shard_block.beacon_chain_ref` is the hash of a block in the (canonical) beacon chain with slot less than or equal to `slot`.
* Verify that `shard_block.beacon_chain_ref` is equal to or a descendant of the `shard_block.beacon_chain_ref` specified in the `ShardBlock` pointed to by `shard_block.parent_root`.
* Let `state` be the state of the beacon chain block referred to by `shard_block.beacon_chain_ref`.

> * Let `persistent_committee` be `[persistent_committee[i] for i in get_persistent_committee(state, shard_block.slot, shard_block.shard_id)`.

Reviewer: missing end bracket

```suggestion
* Let `persistent_committee = get_persistent_committee(state, shard_block.slot, shard_block.shard_id)`.
```

* Assert `verify_bitfield(shard_block.participation_bitfield, len(persistent_committee))`

> * Let `proposer_index = hash(state.randao_mix + int_to_bytes8(shard_block.shard_id) + int_to_bytes8(shard_block.slot)) % len(validators)`. Let `msg` be the `shard_block` but with `shard_block.signature` set to `[0, 0]`. Verify that `bls_verify(pubkey=validators[proposer_index].pubkey, message_hash=hash(msg), signature=shard_block.signature, domain=get_domain(state, shard_block.slot, SHARD_PROPOSER_DOMAIN))` passes.

Reviewer: * Should `state.randao_mix` be `state.current_epoch_seed`? (updated below)
* Don't we need to cast the hash to an int to mod it against `len(validators)`? (updated below)
* `get_domain` takes `epoch` rather than `slot` (updated below)

```suggestion
* Let `proposer_index = bytes_to_int(hash(state.current_epoch_seed + int_to_bytes8(shard_block.shard_id) + int_to_bytes8(shard_block.slot))[0:8]) % len(validators)`. Let `msg` be the `shard_block` but with `shard_block.signature` set to `[0, 0]`. Verify that `bls_verify(pubkey=validators[proposer_index].pubkey, message_hash=hash(msg), signature=shard_block.signature, domain=get_domain(state, slot_to_epoch(shard_block.slot), SHARD_PROPOSER_DOMAIN))` passes.
```

> * Let `group_public_key = bls_aggregate_pubkeys([state.validators[index].pubkey for i, index in enumerate(persistent_committee) if get_bitfield_bit(shard_block.participation_bitfield, i) is True])`. Verify that `bls_verify(pubkey=group_public_key, message_hash=shard_block.parent_root, sig=shard_block.aggregate_signature, domain=get_domain(state, slot, SHARD_ATTESTER_DOMAIN))` passes.

Reviewer: Fix `slot` in `get_domain`

```suggestion
* Let `group_public_key = bls_aggregate_pubkeys([state.validators[index].pubkey for i, index in enumerate(persistent_committee) if get_bitfield_bit(shard_block.participation_bitfield, i) is True])`. Verify that `bls_verify(pubkey=group_public_key, message_hash=shard_block.parent_root, sig=shard_block.aggregate_signature, domain=get_domain(state, slot_to_epoch(shard_block.slot), SHARD_ATTESTER_DOMAIN))` passes.
```

    'beacon_chain_ref': 'bytes32',
    # Merkle root of data
    'data_root': 'bytes32'
    # State root (placeholder for now)
    'state_root': 'bytes32',
    # Block signature
    'signature': 'bytes96',
    # Attestation
    'participation_bitfield': 'bytes',
    'aggregate_signature': 'bytes96',
}
```

## Shard block processing

For a `shard_block` on a shard to be processed by a node, the following conditions must be met:

* The `ShardBlock` pointed to by `shard_block.parent_root` has already been processed and accepted
* The signature for the block from the _proposer_ (see below for definition) of that block is included along with the block in the network message object

To validate a block header on shard `shard_block.shard_id`, compute as follows:

* Verify that `shard_block.beacon_chain_ref` is the hash of a block in the (canonical) beacon chain with slot less than or equal to `slot`.
* Verify that `shard_block.beacon_chain_ref` is equal to or a descendant of the `shard_block.beacon_chain_ref` specified in the `ShardBlock` pointed to by `shard_block.parent_root`.
* Let `state` be the state of the beacon chain block referred to by `shard_block.beacon_chain_ref`.

> * Let `persistent_committee = get_persistent_committee(state, shard_block.slot, shard_block.shard_id)`.

Reviewer: ```suggestion
* Let `persistent_committee = get_persistent_committee(state, shard_block.shard_id, slot_to_epoch(shard_block.slot))`.
```

> * Assert `verify_bitfield(shard_block.participation_bitfield, len(persistent_committee))`

Author: ```suggestion
* Assert `verify_bitfield(shard_block.participation_bitfield, len(persistent_committee))`
* For every `i in range(len(persistent_committee))` where `is_active_validator(state, persistent_committee[i]))` returns `False`, verify that `get_bitfield_bit(shard_block.participation_bitfield, i) == 0`
```

Reviewer: looks good

> * Let `proposer_index = bytes_to_int(hash(state.current_epoch_seed + int_to_bytes8(shard_block.shard_id) + int_to_bytes8(shard_block.slot))[0:8]) % len(validators)`. Let `msg` be the `shard_block` but with `shard_block.signature` set to `[0, 0]`. Verify that `bls_verify(pubkey=validators[proposer_index].pubkey, message_hash=hash(msg), signature=shard_block.signature, domain=get_domain(state, slot_to_epoch(shard_block.slot), SHARD_PROPOSER_DOMAIN))` passes.

Author: ```suggestion
* Let `proposer_index = get_shard_proposer_index(state, state.slot)`. Let `msg` be the `shard_block` but with `shard_block.signature` set to `[0, 0]`. Verify that `bls_verify(pubkey=validators[proposer_index].pubkey, message_hash=hash(msg), signature=shard_block.signature, domain=get_domain(state, slot_to_epoch(shard_block.slot), SHARD_PROPOSER_DOMAIN))` passes.
```

Reviewer: fixed params being passed to `get_shard_proposer_index`. Looks good now.

To validate a block header on shard `shard_block.shard_id`, compute as follows:

* Verify that `shard_block.beacon_chain_ref` is the hash of a block in the (canonical) beacon chain with slot less than or equal to `slot`.
* Verify that `shard_block.beacon_chain_ref` is equal to or a descendant of the `shard_block.beacon_chain_ref` specified in the `ShardBlock` pointed to by `shard_block.parent_root`.
* Let `state` be the state of the beacon chain block referred to by `shard_block.beacon_chain_ref`.
* Let `persistent_committee = get_persistent_committee(state, shard_block.slot, shard_block.shard_id)`.
* Assert `verify_bitfield(shard_block.participation_bitfield, len(persistent_committee))`
* Let `proposer_index = bytes_to_int(hash(state.current_epoch_seed + int_to_bytes8(shard_block.shard_id) + int_to_bytes8(shard_block.slot))[0:8]) % len(validators)`. Let `msg` be the `shard_block` but with `shard_block.signature` set to `[0, 0]`. Verify that `bls_verify(pubkey=validators[proposer_index].pubkey, message_hash=hash(msg), signature=shard_block.signature, domain=get_domain(state, slot_to_epoch(shard_block.slot), SHARD_PROPOSER_DOMAIN))` passes.

> * Let `group_public_key = bls_aggregate_pubkeys([state.validators[index].pubkey for i, index in enumerate(persistent_committee) if get_bitfield_bit(shard_block.participation_bitfield, i) is True])`. Verify that `bls_verify(pubkey=group_public_key, message_hash=shard_block.parent_root, sig=shard_block.aggregate_signature, domain=get_domain(state, slot_to_epoch(shard_block.slot), SHARD_ATTESTER_DOMAIN))` passes.

Reviewer: ```suggestion
* Let `group_public_key = bls_aggregate_pubkeys([state.validators[index].pubkey for i, index in enumerate(persistent_committee) if get_bitfield_bit(shard_block.participation_bitfield, i) is True])`.
* Verify that `bls_verify(pubkey=group_public_key, message_hash=shard_block.parent_root, sig=shard_block.aggregate_signature, domain=get_domain(state, slot_to_epoch(shard_block.slot), SHARD_ATTESTER_DOMAIN))` passes.
```
